unikernel.ml

```diff
@@ -1201,6 +1209,121 @@ struct
+          (reply reqd ~content_type:"text/html"
+             (Dashboard.dashboard_layout ~csrf user
+                ~page_title:(String.capitalize_ascii user.name ^ " | Mollymawk")
+                ~content:(Volume_index.volume_index_layout blocks policy)
+                ~icon:"/images/robur.png" ())
+             ~header_list:[ ("X-MOLLY-CSRF", csrf) ]
+             `OK)
+    | Error err ->
+        Lwt.return
+          (reply reqd ~content_type:"text/html"
+             (Guest_layout.guest_layout ~page_title:"500 | Mollymawk"
+                ~content:(Error_page.error_layout err)
+                ~icon:"/images/robur.png" ())
+             `Internal_server_error)
+
+  let delete_volume json_dict albatross reqd (user : User_model.user) =
+    match Utils.Json.get "block_name" json_dict with
+    | Some (`String block_name) -> (
+        Albatross.query albatross ~domain:user.name ~name:block_name
+          (`Block_cmd `Block_remove)
+        >>= function
+        | Error msg ->
+            Logs.err (fun m -> m "Error querying albatross: %s" msg);
+            Middleware.http_response reqd ~title:"Error"
+              ~data:("Error querying albatross: " ^ msg)
+              `Internal_server_error
+        | Ok (_hdr, res) -> (
+            match Albatross_json.res res with
+            | Ok res ->
+                Middleware.http_response reqd ~title:"Success"
+                  ~data:(Yojson.Basic.to_string res)
+                  `OK
+            | Error (`String res) ->
+                Middleware.http_response reqd ~title:"Error"
+                  ~data:(Yojson.Basic.to_string (`String res))
+                  `Internal_server_error))
+    | _ ->
+        Middleware.http_response reqd ~title:"Error"
+          ~data:"Couldn't find block name in json" `Bad_request
+
+  let create_volume albatross reqd (user : User_model.user) =
+    read_multipart_data reqd >>= fun result ->
+    match result with
+    | Error (`Msg msg) ->
+        Middleware.http_response reqd ~title:"Error"
+          ~data:("Couldn't multipart: " ^ msg)
+          `Bad_request
+    | Ok (m, assoc) -> (
+        let m, _r = to_map ~assoc m in
+        match
+          ( Map.find_opt "block_name" m,
+            Map.find_opt "block_size" m,
+            Map.find_opt "block_data" m,
+            Map.find_opt "block_compressed" m,
+            Map.find_opt "molly_csrf" m )
+        with
+        | ( Some (_, block_name),
+            Some (_, block_size),
+            Some (_, block_data),
+            Some (_, block_compressed),
+            Some (_, form_csrf_token) ) ->
+            (let now = Ptime.v (P.now_d_ps ()) in
+             Middleware.csrf_verification user now form_csrf_token (fun reqd ->
+                 Albatross.query albatross ~domain:user.name ~name:block_name
+                   (`Block_cmd
+                     (`Block_add
+                       ( int_of_string block_size,
+                         bool_of_string block_compressed,
```

Comment:
here we should take care of errors / exceptions: since block_size is user input, as is block_compressed, we need to be more careful.

I also think we should use proper json types -- `` `Int `` and  `` `Bool `` -- for these fields, if possible.

Reply:
@hannesm  We are using multipart/form-data for this post request. perhaps we can jsonify the fields which aren't files and then we can use json types for them. wdyt?

Reply:
yes, that sounds good to me :)

Reply:
fixed in https://github.com/robur-coop/mollymawk/pull/90/commits/ab22fe969b288937af2b891326a515e5c1025e1e